### ISSUE TITLE
Wrong behavior where two foreign keys targeting the same table

### ISSUE THREAD

**Issue by rouxjeremy** [ignasbernotas/dexie-relationships#14]
Hi,

I tried to declare two foreign key targeting the same table, but I got the following issue : 

```
stores = {
    'albums': 'id',
    'bands': 'id, firstAlbumId -> albums.id, lastAlbumId -> albums.id'
};

db.bands.with({
    firstAlbum: 'firstAlbumId',
    lastAlbum: 'lastAlbumId'
}).then(function (bands) {
    bands.forEach(function (band) {
        console.log('band : ' + band.name)
        console.log('first :' + band.fisrtAlbum);
        console.log('last : ' + band.lastAlbum);
        // in this case, fisrt album is always undefined
    });
});

db.bands.with({
    lastAlbum: 'lastAlbumId',
    firstAlbum: 'firstAlbumId'
}).then(function (bands) {
    bands.forEach(function (band) {
        console.log('band : ' + band.name)
        console.log('first :' + band.fisrtAlbum);
        console.log('last : ' + band.lastAlbum);
        // in this case, last album is always undefined
    });
});
```

The second relationship overwrite the first, because foreign tables are declared in an object.
Then, I fixed it locally by declaring usableForeignTables as an array of objects containing tableName as property in addition of existing properties : it works well.

You can see below the new "with" method for comparison (sorry, I only edited the dist version).

```
  /**
   * Iterate through all items and collect related records
   *
   * @param relationships
   *
   * @returns {Dexie.Promise}
   */
  db.Collection.prototype.with = function (relationships) {
    var this$1 = this;

    var baseTable = this._ctx.table.name;
    var databaseTables = db._allTables;

    // this holds tables that have foreign keys pointing at the current table
    var usableForeignTables = [];

    // validate target tables and add them into our usable tables object
    Object.keys(relationships).forEach(function (column) {
      var tableOrIndex = relationships[column];
      var matchingIndex = this$1._ctx.table.schema.idxByName[tableOrIndex];

      if (matchingIndex && matchingIndex.hasOwnProperty('foreignKey')) {
        var index = matchingIndex;
        usableForeignTables.push({
          column: column,
          index: index.foreignKey.targetIndex,
          tableName: index.foreignKey.targetTable,
          targetIndex: index.foreignKey.index,
          oneToOne: true
        });
      } else {
        var table = tableOrIndex;

        if (!databaseTables.hasOwnProperty(table)) {
          throw new Error('Relationship table ' + table + ' doesn\'t exist.');
        }

        if (!databaseTables[table].schema.hasOwnProperty('foreignKeys')) {
          throw new Error('Relationship table ' + table + ' doesn\'t have foreign keys set.');
        }

        // remove the foreign keys that don't link to the base table
        var columns = databaseTables[table].schema.foreignKeys.filter(function (column) { return column.targetTable === baseTable; });

        if (columns.length > 0) {
          usableForeignTables[table] = {
            column: column,
            index: columns[0].index,
            tableName: table,
            targetIndex: columns[0].targetIndex
          };
        }
      }
    });

    return this.toArray().then(function (rows) {
      //
      // Extract the mix of all related keys in all rows
      //
      var queries = usableForeignTables
        .map(function (foreignTable) {
          // For each foreign table, query all items that any row refers to
          var tableName = foreignTable.tableName;
          var allRelatedKeys = rows
            .map(function (row) { return row[foreignTable.targetIndex]; })
            .filter(isIndexableType);

          // Build the Collection to retrieve all related items
          return databaseTables[tableName]
              .where(foreignTable.index)
              .anyOf(allRelatedKeys);
        });

      // Execute queries in parallell
      var queryPromises = queries.map(function (query) { return query.toArray(); });

      //
      // Await all results and then try mapping each response
      // with its corresponding row and attach related items onto each row
      //
      return Promise.all(queryPromises).then(function (queryResults) {
        usableForeignTables.forEach(function (foreignTable, idx) {
          var tableName = foreignTable.tableName;
          var result = queryResults[idx];
          var targetIndex = foreignTable.targetIndex;
          var foreignIndex = foreignTable.index;
          var column = foreignTable.column;

          // Create a lookup by targetIndex (normally 'id')
          // and set the column onto the target
          var lookup = {};
          rows.forEach(function (row) {
            lookup[row[targetIndex]] = row;
          });

          // Populate column on each row
          result.forEach(function (record) {
            var foreignKey = record[foreignIndex];
            var row = lookup[foreignKey];
            if (!row) {
              throw new Error(
                "Could not lookup foreign key where " +
                tableName + "." + foreignIndex + " == " + baseTable + "." + column + ". " +
                "The content of the failing key was: " + (JSON.stringify(foreignKey)) + ".");
            }

            if (foreignTable.oneToOne || !row.hasOwnProperty(column)) {
              // Set it as a non-enumerable property so that the object can be safely put back
              // to indexeddb without storing relations redundantly (IndexedDB will only store "own non-
              // enumerable properties")
              Object.defineProperty(row, column, {
                value: foreignTable.oneToOne ? record : [record],
                enumerable: false,
                configurable: true,
                writable: true
              });
            } else if (!foreignTable.oneToOne) {
              row[column].push(record);
            }
          });
        });
      }).then(function () { return rows; });
    });
  };
```

By the way, thank you for this usefull plugin, I just started using it in my company and except this issue, it works well !

I hope it will be fixed soon in order to use it from npm :)

**Comment by dfahlander** [comment]
@nybblr, do you think this is the same issue you resolved?

**Comment by nybblr** [comment]
@dfahlander It's definitely from the same hotspot (`var lookup` to the end), but I don't think my PR would fix this since the issue occurs earlier when `usableForeignTables` is declared.

**Comment by dfahlander** [comment]
@rouxjeremy Tried to apply your fix in a branch at https://github.com/dfahlander/dexie-relationships/tree/issue14. Problem is that some browser-based unit tests fails.

If you (or someone else) have a change to look at it, do the following:

1. Clone https://github.com/dfahlander/dexie-relationships
2. checkout branch issue14
3. npm test
4. npm install -g http-server
5. http-server -c-1
6. Open web browser to http://localhost:8080/test/browser/testrunner.html

Now the map files are there so you can set break points from the source code in chrome debugger.

**Comment by CraigCav** [comment]
@dfahlander I think you were pretty close with your branch.

I've [tweaked it](https://github.com/CraigCav/dexie-relationships/commit/a8b61f21b125ba8702b036dd9b76600fc2391457) slightly (there was still one location that was treating `usableForeignTables` as a dictionary, rather than an array), and now the browser unit tests all pass.

Running this change locally seems to resolve the issue.

Here's my branch with the necessary change: https://github.com/CraigCav/dexie-relationships/tree/issue14

**Comment by dfahlander** [comment]
Thanks @CraigCav for the final fix. I think this module would benefit a great deal if it would be rewritten in typescript.

**Comment by xVinCe** [comment]
Hello,

There is the same problem the other way around, when you try to populate 2 arrays leading to the same table. Example (not very meaningful sorry :) ):

`stores = {
    'albums': 'id, primaryGenreId -> genres.id, secondaryGenreId -> genres.id',
    'genres': 'id'
};

db.genres.with({
    primaryAlbums: 'albums',
    secondaryAlbums: 'albums'
}).then(function (genres) {
    genres.forEach(function (genre) {
        console.log('genre : ' + genre.name)
        console.log('primary :', genre.primaryAlbums); // OK
        console.log('last : ', genre.secondaryAlbums); // KO => gets primary albums twice
    });
});`

Of course, Dexie-relationship has on way of knowing which key should be used, as we only mention the table name.

So, I have fixed it this way:
Add the key after the table in the with:
`db.genres.with({
    primaryAlbums: 'albums.primaryGenreId',
    secondaryAlbums: 'albums.primaryGenreId'
})`

And then I have patched the lib (see attached file). It doesn't break the existing behavior, and adds the possibility to specify the primary key on the related table.
[dexie-relationships+1.2.10.patch.zip](https://github.com/ignasbernotas/dexie-relationships/files/3090491/dexie-relationships%2B1.2.10.patch.zip)

(My patch also contains the fix for https://github.com/ignasbernotas/dexie-relationships/issues/28)
